File: currency-rate/src/main/java/com/nao4j/subtrol/currencyrate/service/CurrencyServiceImpl.java
package com.nao4j.subtrol.currencyrate.service;

import com.nao4j.subtrol.currencyrate.document.CurrencyRates;
import com.nao4j.subtrol.currencyrate.dto.CurrencyRate;
import com.nao4j.subtrol.currencyrate.provider.CurrencyRatesProvider;
import com.nao4j.subtrol.currencyrate.repository.CurrencyRatesRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;

import static java.math.BigDecimal.ONE;
import static java.math.RoundingMode.HALF_UP;
import static java.util.stream.Collectors.toCollection;

@Service
@RequiredArgsConstructor
public class CurrencyServiceImpl implements CurrencyService {

    private final CurrencyRatesRepository cacheRepository;
    private final CurrencyRatesProvider currencyRatesProvider;

    @Override
    public Set<String> getAll() {
        return cacheRepository.findAll().stream()
                .flatMap(currencyRates -> currencyRates.getRates().keySet().stream())
                .collect(toCollection(TreeSet::new));
    }

    @Override
    public CurrencyRate getRate(final String source, final String target) {
        final var fixerData = cacheRepository.findTopByOrderByTimestampDesc()
                .orElseThrow(IllegalArgumentException::new);
        if (source.equals(fixerData.getBase())) {
            return new CurrencyRate(source, target, fixerData.getRates().get(target));
        } else if (target.equals(fixerData.getBase())) {
            return new CurrencyRate(source, target, ONE.divide(fixerData.getRates().get(target), 6, HALF_UP));
        }
        final var sourceRate = fixerData.getRates().get(source);
        final var targetRate = fixerData.getRates().get(target);
        return new CurrencyRate(source, target, targetRate.divide(sourceRate, 6, HALF_UP));
    }

    @Override
    public Optional<CurrencyRates> loadData() {
        return currencyRatesProvider.latest().map(cacheRepository::save);
    }

    @Override
    public long deleteOldest(final LocalDate date) {
        return cacheRepository.deleteAllByDateBefore(date);
    }

}

File: src/main/kotlin/com/nao4j/subtrol/service/CurrencyServiceImpl.kt
package com.nao4j.subtrol.service

import com.nao4j.subtrol.model.CurrencyRate
import com.nao4j.subtrol.model.CurrencyRates
import com.nao4j.subtrol.repository.FixerApiRepository
import com.nao4j.subtrol.repository.FixerCacheRepository
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.math.BigDecimal.ONE
import java.math.RoundingMode
import java.time.LocalDate

@Service
class CurrencyServiceImpl(
        val cacheRepository: FixerCacheRepository,
        val api: FixerApiRepository
): CurrencyService {

    override fun getAll(): Set<String> {
        return api.latest()?.rates?.keys ?: emptySet()
    }

    override fun getRate(source: String, target: String): CurrencyRate {
        val fixerData = cacheRepository.findTopByOrderByTimestampDesc()
        return when {
            (source == fixerData.base) ->
                CurrencyRate(source, target, fixerData.rates[target]!!)
            (target == fixerData.base) ->
                CurrencyRate(
                        source,
                        target,
                        ONE.divide(fixerData.rates[target]!!, 6, RoundingMode.HALF_UP)
                )
            else -> {
                val sourceRate = fixerData.rates[source]!!
                val targetRate = fixerData.rates[target]!!
                CurrencyRate(source, target, targetRate.divide(sourceRate, 6, RoundingMode.HALF_UP))
            }
        }
    }

    @Scheduled(cron = "0 0 3 1/1 * ?")
    override fun loadData(): CurrencyRates? {
        val fixerData = api.latest()
        if (fixerData != null) {
            return cacheRepository.save(fixerData)
        }
        return null
    }

    override fun deleteOldest(date: LocalDate): Long = cacheRepository.deleteAllByDateBefore(date)

}
